src/main/java/com/google/javascript/cl2dts/DeclarationGenerator.java

```diff
@@ -148,37 +157,54 @@ public String produceDts(Compiler compiler) {
           isDefault = true;
         }
       }
-      emitNoSpace("declare namespace ");
-      emitNoSpace(INTERNAL_NAMESPACE);
-      if (!namespace.isEmpty()) {
-        emitNoSpace("." + namespace);
-      }
-      emitNoSpace(" {");
-      indent();
-      emitBreak();
-      if (symbol.getType() != null) {
-        walkScope(symbol, namespace, isDefault);
-      } else {
-        // JSCompiler treats "foo.x" as one variable name, so collect all provides that start with
-        // $provide + ".".
-        String prefix = provide + ".";
-        for (TypedVar other : topScope.getAllSymbols()) {
-          String otherName = other.getName();
-          if (otherName.startsWith(prefix) && other.getType() != null
-              && !other.getType().isFunctionPrototypeType()) {
-            walkScope(other, namespace, false);
-          }
-        }
-      }
-      unindent();
-      emit("}");
-      emitBreak();
+      currentNamespace = namespace;
```

Comment:
Okay, fresh with a bit of extra sleep the refactoring was no problem.
Essentially I've introduced an inner class to replace the walkScope method, and the lifecycle of this class is one-per-top-level-symbol. So the namespace and the top level symbol can be state of this class, and it doesn't require changing *every* visit* method to accept a namespace parameter and pass it down.